File: 101-109_small_problems/easy_7/swap_case.rb
# Swap Case
# Write a method that takes a string as an argument and returns a new string in which every uppercase letter is replaced by its lowercase version, and every lowercase letter by its uppercase version. All other characters should be unchanged.

# You may not use String#swapcase; write your own version of this method.

# Input: A string
# Output: A new string
# Mental model: Given a string, return a new string in which each letter has its' case swapped. Don't change any other characters.

# Example:

# swapcase('CamelCase') == 'cAMELcASE'
# swapcase('Tonight on XYZ-TV') == 'tONIGHT ON xyz-tv'

# Data structure
# Input and output can be strings.

# Algorithm
# Split the string into characters
# Loop over the characters
#   If character is uppercase
#     make character lowercase
#   Else
#     make character uppercase

# Code

def swapcase(str)
  swapped_case = str.chars.map do |char|
    char == char.upcase ? char.downcase : char.upcase
  end

  swapped_case.join
end

p swapcase('CamelCase') == 'cAMELcASE'
p swapcase('Tonight on XYZ-TV') == 'tONIGHT ON xyz-tv'

File: exercises/js210_small_problems/string_and_text/swapCase.js
// Swap Case
// Write a function that takes a string as an argument, and returns that string with every lowercase letter changed to uppercase and every uppercase letter changed to lowercase. Leave all other characters unchanged.
//
// Input: A string
// Output: A string
//
// Rules
// - lowercase characters become uppercase
// - uppercase characters become lowercase
// - all other characters are unchanged
//
// Transformation problem
//
// Data Structures
// turn string into an array to transform each character
//
// Algorithm
// 0. Split string into an array
// 1. Loop through array
// 2.   Uppercase character if character is currently lowercase
// 3.   Lowercase character if character is currently uppercase
// 4.   Return character if character is not alphabetic
// 5. Join array into a string
// 6. Return string

// Code
function swapCase(str) {
  return str.split('')
            .map(function (char) {
              if (/[a-z]/.test(char)) {
                return char.toUpperCase();
              } else if (/[A-Z]/.test(char)) {
                return char.toLowerCase();
              } else {
                return char;
              }
            })
            .join('');
}
// Examples:

console.log(swapCase('CamelCase'));              // "cAMELcASE"
console.log(swapCase('Tonight on XYZ-TV'));      // "tONIGHT ON xyz-tv"
